go/fixchain/hash.go

```diff
@@ -0,0 +1,60 @@
+package fixchain
+
+import (
+	"crypto/sha256"
+	"sort"
+
+	"github.com/google/certificate-transparency/go/x509"
+)
+
+const hashSize = sha256.Size
+
+var newHash = sha256.New
+
+func hash(c *x509.Certificate) (hash [hashSize]byte) {
+	copy(hash[:], newHash().Sum(c.Raw))
+	return
+}
+
+func hashChain(ch []*x509.Certificate) (hash [hashSize]byte) {
+	h := newHash()
+	for _, c := range ch {
+		h.Write(newHash().Sum(c.Raw))
+	}
+	copy(hash[:], h.Sum(nil))
+	return
+}
+
+func hashBag(chain []*x509.Certificate) [hashSize]byte {
```

Comment:
In the original code, hashBag was for caching in the Fixer.  It was used to hash a chain that had had every cert in it queued for fixing and logging.  I removed this because it didn't make sense for the new Fixer, but it does make sense for the FixAndLog.  I have just added caching to the FixAndLog using hashBag to calculate the chain hashes.  In our use case the chains should always be deduped, but I have not put this functionality in hashBag because in the one place it is called the chain is deduped right before calling hashBag (and the deduped chain is then used after calling hashBag too).  In general (and in the tests) it does not assume that hashBag is only for deduped chains, and hashing two chains containing the same certs, but where one has some additional duplicates, does not produce the same hash.